ethexe/service/src/fast_sync.rs

```diff
@@ -0,0 +1,604 @@
+                }
+            }
+
+            let header = OnChainStorage::block_header(db, block)
+                .ok_or_else(|| anyhow!("header not found for synced block {block}"))?;
+            let parent = header.parent_hash;
+            block = parent;
+        }
+
+        if let Some(latest_block) = latest_block {
+            // recover data we haven't seen in events by the latest computed block
+            if let Some((computed_block, _computed_header)) = db.latest_computed_block() {
+                let computed_program_states = db
+                    .block_program_states(computed_block)
+                    .context("program states of latest computed block not found")?;
+                for (program_id, state) in computed_program_states {
+                    program_states.entry(program_id).or_insert(state);
+                }
+            }
+
+            #[cfg(debug_assertions)]
+            if let Some(previous_block) = previous_block {
+                let latest_block_header = OnChainStorage::block_header(db, latest_block)
+                    .expect("observer must fulfill database");
+                let previous_block_header = OnChainStorage::block_header(db, previous_block)
+                    .expect("observer must fulfill database");
+                assert!(previous_block_header.height < latest_block_header.height);
+            }
+        }
+
+        Ok(Self {
+            program_states,
+            program_code_ids,
+            code_ids,
+            latest_block,
+            previous_block,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum RequestMetadata<'a> {
+    ProgramState {
+        program_ids: &'a HashSet<ActorId>,
+    },
+    MemoryPages,
+    MemoryPagesRegion,
+    Waitlist {
+        program_ids: &'a HashSet<ActorId>,
+    },
+    Mailbox {
+        program_ids: &'a HashSet<ActorId>,
+    },
+    UserMailbox {
+        program_ids: &'a HashSet<ActorId>,
+        user_id: ActorId,
+    },
+    DispatchStash {
+        program_ids: &'a HashSet<ActorId>,
+    },
+    /// Any data we only insert into the database.
+    Data,
+}
+
+impl RequestMetadata<'_> {
+    /// Simple request metadata means that there is no need to process its content
+    /// if it's already in the database because it has no fields that affect data flow.
+    ///
+    /// For example, `ProgramState` has a list of program IDs,
+    /// so we need to process the whole request tree up to `Waitlist`, `UserMailbox` and others
+    /// because they can change `ScheduleRestorer`.
+    fn is_simple(self) -> bool {
+        matches!(
+            self,
+            RequestMetadata::MemoryPages
+                | RequestMetadata::MemoryPagesRegion
+                | RequestMetadata::Data
+        )
+    }
+}
+
+#[derive(Debug, Default)]
+struct RequestManager<'a> {
+    /// Total completed requests
+    total_completed_requests: u64,
+    /// Total pending requests
+    total_pending_requests: u64,
```

Comment:
total_network_requests. Using name total_pending_requests looks like it's just a len of pending _requests

Reply:
> it's just a len of pending _requests

That's why the name is prefixed with `total_`

Reply:
still not clear for me, but now you use it only for debug checks and stats, so not important